src/main/java/net/minecraftforge/gradle/user/patch/UserPatchBasePlugin.java

```diff
@@ -130,10 +146,118 @@ public void execute(Object arg0) {
             }
 
         });
+
+        // Make sure to clear build cache before compiling/processing classes/resources.
+        // Required for proper inflation/token replacement when building
+        this.project.getTasks().getByName("sourceMainJava").dependsOn(this.makeTask("clearBuildCache", ClearBuildTask.class));
+        this.project.getTasks().getByName("processResources").dependsOn(this.makeTask("clearResourcesCache", ClearResourcesTask.class));
+
+        // Setting custom archive classifiers for some reason causes Gradle to clean
+        // build output right before reobf task is executed. This should help prevent it
+        ReobfTask reobf = (ReobfTask) this.project.getTasks().getByName("reobf");
+        reobf.getInputs().dir(this.project.getBuildDir());
+        reobf.getOutputs().dir(this.project.getBuildDir());
+
+        Task test = this.project.getTasks().getByName("test");
+        test.getInputs().dir(this.project.getBuildDir());
+        test.getOutputs().dir(this.project.getBuildDir());
+
+        // Check if user enabled our incredibly very shenanigans and make them happen if so
+        if (this.areGrimoireShenanigansEnabled()) {
+            this.applyGrimoireShenanigans();
+        }
+
+        // Inform the user whether or not Grimoire shenanigans are enabled.
+        // Doing this after evaluate so that banner gets printed first
+        this.project.afterEvaluate(new Action() {
+
+            @Override
+            public void execute(Object arg) {
+                UserPatchBasePlugin.this.project.getLogger().lifecycle("Grimoire shenanigans " +
+                        (UserPatchBasePlugin.this.areGrimoireShenanigansEnabled() ? "enabled" : "disabled"));
+
+                if (UserPatchBasePlugin.this.areGrimoireShenanigansEnabled()) {
+                    UserPatchBasePlugin.this.project.getLogger().lifecycle("Using Mixin refmap name: " +
+                            UserPatchBasePlugin.this.getMixinRefmapName());
+                }
+            }
+
+        });
+
+    }
+
+    @SuppressWarnings({"unchecked", "rawtypes"})
+    protected void applyGrimoireShenanigans() {
+        // Add important repositories for Mixin annotation processor
+        this.addMavenRepo(this.project, "spongepowered-repo", "https://repo.spongepowered.org/maven/");
+        this.project.getRepositories().mavenCentral();
+
+        // Make custom task for copying srg-files into build/srgs directory
+        CopySrgsTask copyTask = this.makeTask("copySrgs", CopySrgsTask.class);
+        copyTask.init(this);
+        copyTask.dependsOn("genSrgs");
+
+        JavaCompile compileJava = (JavaCompile) this.project.getTasks().getByName("compileJava");
+        compileJava.dependsOn(copyTask);
+
+        // Add annotation processor for... annotation processing, I suppose
+        this.project.getDependencies().add("annotationProcessor", "org.spongepowered:mixin:0.7.11-SNAPSHOT");
```

Comment:
Maybe later add a way to change sponge mixin version.

Reply:
It would only make sense if Grimoire itself did support varying versions of Mixin, which it doesn't. We can probably try to support any in-dev Mixin setups, but going forward I assume it will be more efficient to focus our effort on Grimoire support alone.